driver-core/src/main/com/mongodb/internal/operation/CommandOperationHelper.java

```diff
@@ -227,42 +206,44 @@ public T call(final ConnectionSource source) {
 
             if (!shouldAttemptToRetryRead(retryReads, e)) {
                 if (retryReads) {
-                    logUnableToRetry(command.getFirstKey(), e);
+                    logUnableToRetry(Objects.requireNonNull(command).getFirstKey(), e);
                 }
                 throw exception;
             }
         } finally {
             connection.release();
         }
 
-        final MongoException originalException = exception;
-        return withReleasableConnection(binding, originalException, new CallableWithConnectionAndSource<T>() {
-            @Override
-            public T call(final ConnectionSource source, final Connection connection) {
-                try {
-                    if (!canRetryRead(source.getServerDescription(), connection.getDescription(), binding.getSessionContext())) {
-                        throw originalException;
-                    }
-                    BsonDocument retryCommand = commandCreator.create(source.getServerDescription(), connection.getDescription());
-                    logRetryExecute(retryCommand.getFirstKey(), originalException);
-                    return executeCommand(database, retryCommand, decoder, source, connection, binding.getReadPreference(), transformer,
-                            binding.getSessionContext(), source.getServerApi());
-                } finally {
-                    connection.release();
+        return retryCommand(binding, database, commandCreator, decoder, transformer, exception);
+    }
+
+    private static <D, T> T retryCommand(final ReadBinding binding, final String database, final CommandCreator commandCreator,
+                                         final Decoder<D> decoder, final CommandReadTransformer<D, T> transformer,
+                                         final MongoException originalException) {
+        return withReleasableConnection(binding, originalException, (source, connection) -> {
```

Comment:
Suggestion: I think in the future we should look to clean up these loan methods `withConnection` and `withReleasableConnection`.  So that they are inline with `withReadConnectionSource` and handle the try/finally releasing.  Reduces the boilerplate and risk of not releasing the connection.

Reply:
Ack.